### PR TITLE
Docker compose

### PR DESCRIPTION
**Pull request by Codesleuth**
This change adds a `Dockerfile` and `docker-compose.yml` file to be able to run the API as either a composed set with database, or as a standalone container.

To get the compose file to work, I had to add retrying of connection attempts when mongoose fails to connect (slow MongoDB startup in docker).

I have also updated all dependencies and fixed all new lint issues recommended in the `tslint:recommended` set.

### CODE REVIEW

**Comment by adamwitko** [comment]
🤔 🌮 👍

File: docker-compose.yml [hunk 1]
```diff
@@ -1,6 +1,16 @@
...
+    image: mongo:3.4.3
     ports:
       - 27017:27017
+    command: --smallfiles
+  api:
+    build: .
+    depends_on:
+      - mongo
+    environment:
+      - MONGODB_URL=mongodb://mongo:27017/local?connectTimeoutMS=30000
```
Comment (adamwitko) [inline docker-compose.yml]: Perhaps don't use 27017 for us unfortunate bunch that have this bound to 27017 on our host?

Comment (Codesleuth) [inline docker-compose.yml]: This should be isolated in the default network in docker. Did you have an issue?

Comment (Codesleuth) [inline docker-compose.yml]: Oh right - wait - I've exposed the MongoDB port in the service above it. It's unnecessary I think, unless you want to be able to connect and have a look at it.

Comment (Codesleuth) [inline docker-compose.yml]: @adamwitko should be fixed. Can you try?

**Pull request merged**